### DISCUSSION TITLE
is passing callback function to atom setter okay?

### DISCUSSION THREAD

**Discussion by wootra** [pmndrs/jotai#1650] [Q&A] [answered]
Hello, team!
I'd like to ask a tedious question about what is recommended pattern in this situation considering of jotai version upgrade in the future. 

I have a code like this:
```
const updateValueAtom = atom(null, (get, set, {obj, callback})=>{
  ...do something with obj
  const newVal = { ...result from above process };
  callback && queueMicrotask(()=>callback(newVal));
});

const useUpdateValue = ()=>useSetAtom(updateValueAtom);

const MyComponent = ()=>{
  const [, dispatch] = useAtom(someAtomWithReducer);
  const updateValue = useUpdateValue();
  useEffect(()=>{
    updateValue({val: 1}, (newVal)=>dispatch(someAction(newVal)));
  }, []);
```

the reason I used this callback is to call dispatch at the moment when updateValueAtom logic is finished as well as taking advantage of its result value. 
I want to use dispatch function as it is because it can be any other function that I don't want to modify. 
I added queueMicrotask to ensure this callback function is called after jotai setter trigger rendering. 

Since I don't want to call any render-triggering function in the other useEffect like `useEffect(()=>doSomething, [newVal])`, sending callback was the easiest way that I could use. 
But if there is a better and modest way of doing this work, it'll be nice because passing callback makes the logic a little complex and embedded. 

Any suggestion will be appreciated. Thanks!!

**Comment by dai-shi** [comment, marked as answer]
1. passing callback function to atom setter is okay (in both current version and future version.)
2. calling atom setter in useEffect is only unavoidable practice when we need to bridge react states/props and jotai states.
3. queueMicrotask doesn't help ensure waiting for next rendering. queueMicrotask shouldn't be required to ensure the value is updated.
4. passing callback function could be avoided if you have full control with atoms. we usually prefer putting logic in atoms not in components.

**Reply by wootra** [reply]
thanks for the quick answer!

**Reply by wootra** [reply]
Hello! 
Sorry for bothering.
While I was thinking this is clear, my coworker who read this answer interpret differently from what I did. 
Would you mind giving me more clarifications?
1. he think passing callback "should" be avoided if possible, while I think that's totally fine.
2. he think useEffect for the initializing atom should be avoided even though I have to call dispatch function right after atom value is changed. 

For more context, we should call a service for setting the initial state of few atoms, so atom setter was used. 
After setting atom values, we should call one more dispatch functions to load a list of data.
in the dispatch function, we set loading state atom, call a service, and set another atom for the list from the service, and unset loading state atom. 

while I have choice to put all the logic in the setter atom, I prefer to make the atoms smaller and modularized, I prefer passing callback function for the following process ( set loading, call service, unset loading ). 

I want to know if passing callback as atom setter for finished event is really bad design. 
@dai-shi

**Reply by dai-shi** [reply]
1. passing callback is valid. if it's intentional, it's not bad design. I prefer to put logic more in atoms, so I probably wouldn't design atoms to receive callbacks if possible. It's more or less preference on atom model.
2. useEffect to initialize atom should be avoided whenever possible, because useEffect delays to run the effect function. It is a valid usage to sync with props/states from other libs, that can't be used as atoms.

Hope it clarifies.

**Reply by wootra** [reply]
Thank you so much! that helped a lot!
